### PR TITLE
[ROCm] [Bugfix] Fix DeepSeek V4 Functionality and Accuracy

### PR DESCRIPTION
**Pull request by tjtanaa**
## Purpose

This PR addresses the following issues

1. Functionality broken after this PR https://github.com/vllm-project/vllm/pull/41263 (fixed by removing `x = self.ffn_norm(x)`)
2. Fix the DeepSeek V4 accuracy degradation at high concurrency.

I have added self-review comments to explain the changes.

### Point 2
It is a compounded issue.
1. AITER MHC has accuracy issue when number of tokens is large. New AITER version is coming soon, so the code path is kept documented so that we can easily reenable it in vLLM once we upgrade the AITER version. For now we are falling back to torch implementation.
2. The prefill mqa topk buffer is corrupted by incorrect logic. It is fixed by calling the `torch.ops._C.top_k_per_row_prefill` .

Other changes are code cleaning and also invoking faster kernels e.g. replace `_topk_indices_torch` with `torch.ops._C.top_k_per_row_decode` . 

## Test Plan

lmeval score of `deepseek-ai/DeepSeek-V4-Pro` and `deepseek-ai/DeepSeek-V4-Flash`

Server command
```
#!/bin/bash
max_num_seqs=256
max_model_len=1048576
tensor_parallel_size=8
rm -rf ~/.cache/vllm

export VLLM_ROCM_USE_AITER=1
vllm serve deepseek-ai/DeepSeek-V4-Pro \
    --host localhost \
    --port 8001 \
    --kv-cache-dtype fp8 \
    --block-size 256 \
    --tensor-parallel-size "${tensor_parallel_size}" \
    --max-model-len "${max_model_len}" \
    --tensor-parallel-size 8 \
    --max-num-seqs 256 \
    --distributed-executor-backend mp \
    --trust-remote-code \
    --gpu-memory-utilization 0.6 \
    --moe-backend triton_unfused \
    --tokenizer-mode deepseek_v4 \
    --reasoning-parser deepseek_v4 \
    --async-scheduling \
    --enable-prefix-caching \
    --compilation-config '{"mode":3,"cudagraph_mode": "FULL_AND_PIECEWISE"}' \
| tee deepseekv4_graph_mhctriton_block256_topkindicesvllmop_syncmain.log
```

lm eval command:

```bash
#!/bin/bash

SCRIPT_DIR=$(cd -- "$(dirname -- "${BASH_SOURCE[0]}")" && pwd)
LM_EVAL_PYTHON=${LM_EVAL_PYTHON:-"${SCRIPT_DIR}/hfvenv/bin/python"}
MODEL=deepseek-ai/DeepSeek-V4-Pro
"${LM_EVAL_PYTHON}" -m lm_eval --model local-completions --model_args model=$MODEL,base_url=http://0.0.0.0:8001/v1/completions,num_concurrent=256,max_retries=10,max_gen_toks=2048,max_length=1048576,timeout=60000 --batch_size auto --tasks gsm8k --num_fewshot 8 \
  --output_path ./results_graph_numshot8 \
  --log_samples \
| tee lmeval_graph_numshot8.log
```

## Test Result

deepseek-ai/DeepSeek-V4-Pro

```
local-completions ({'model': 'deepseek-ai/DeepSeek-V4-Pro', 'base_url': 'http://0.0.0.0:8001/v1/completions', 'num_concurrent': 256, 'max_retries': 10, 'max_gen_toks': 2048, 'max_length': 1048576, 'timeout': 60000}), gen_kwargs: ({}), limit: None, num_fewshot: 8, batch_size: auto
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     8|exact_match|_  |0.9553|_  |0.0057|
|     |       |strict-match    |     8|exact_match|_  |0.9560|_  |0.0056|
```

deepseek-ai/DeepSeek-V4-Flash

```
local-completions ({'model': 'deepseek-ai/DeepSeek-V4-Flash', 'base_url': 'http://0.0.0.0:8001/v1/completions', 'num_concurrent': 256, 'max_retries': 10, 'max_gen_toks': 2048, 'max_length': 1048576, 'timeout': 60000}), gen_kwargs: ({}), limit: None, num_fewshot: 8, batch_size: auto
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     8|exact_match|_  |0.9492|_  | 0.006|
|     |       |strict-match    |     8|exact_match|_  |0.9500|_  | 0.006|
```

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/v1/attention/ops/rocm_aiter_mla_sparse.py [hunk 1]
```diff
@@ -541,7 +541,11 @@ def rocm_fp8_mqa_logits(
         return fp8_mqa_logits_torch(q, kv, weights, cu_seqlen_ks, cu_seqlen_ke)
 
 
-def _topk_indices_torch(logits: torch.Tensor, topk_tokens: int) -> torch.Tensor:
+def _topk_indices_torch(
```
Comment (tjtanaa) [inline vllm/v1/attention/ops/rocm_aiter_mla_sparse.py:545]: This is a fixed to the reference `_topk_indices_torch` so that it is aligned with the `torch.ops._C.top_k_per_row_prefill` implementation.

File: vllm/v1/attention/ops/rocm_aiter_mla_sparse.py [hunk 2]
```diff
@@ -562,64 +572,6 @@ def _topk_indices_torch(logits: torch.Tensor, topk_tokens: int) -> torch.Tensor:
...
-    cu_seqlen_ks: torch.Tensor,
-    cu_seqlen_ke: torch.Tensor,
-) -> None:
-    """Top-k indices for the prefill path.
-
-    Writes ``logits.shape[0]`` rows into ``topk_out``; caller must size the
-    view accordingly.
-    """
-    if topk_tokens in _TOPK_FAST_PATH_VALUES:
-        torch.ops._C.top_k_per_row_prefill(
```
Comment (tjtanaa) [inline vllm/v1/attention/ops/rocm_aiter_mla_sparse.py:583]: Since `torch.ops._C.top_k_per_row_prefill` supports general values of `topk_tokens`, we can call it all the time. Not just when `topk_tokens=2048`

File: vllm/v1/attention/ops/rocm_aiter_mla_sparse.py [hunk 3]
```diff
@@ -634,21 +586,12 @@ def rocm_aiter_sparse_attn_indexer_fake(
     max_model_len: int,
     total_seq_lens: int,
     topk_indices_buffer: torch.Tensor | None,
+    skip_k_cache_insert: bool = False,
```
Comment (tjtanaa) [inline vllm/v1/attention/ops/rocm_aiter_mla_sparse.py:590]: I have cleaned up the function interface following the CUDA's `sparse_attn_indexer_fake` and `sparse_attn_indexer` implementation.

File: vllm/v1/attention/ops/rocm_aiter_mla_sparse.py [hunk 4]
```diff
@@ -836,39 +788,6 @@ def rocm_aiter_sparse_attn_indexer_native(
     return topk_indices_buffer
 
 
-def rocm_aiter_sparse_attn_indexer(
```
Comment (tjtanaa) [inline vllm/v1/attention/ops/rocm_aiter_mla_sparse.py:841]: Chores: We don't need another layer of wrapper.

File: vllm/model_executor/layers/sparse_attn_indexer.py [hunk 5]
```diff
@@ -539,5 +518,9 @@ def forward_hip(
                 self.max_model_len,
                 self.max_total_seq_len,
                 self.topk_indices_buffer,
+                skip_k_cache_insert=self.skip_k_cache_insert,
             )
-        raise RuntimeError("Sparse attention indexer ROCm path could not be selected.")
+        raise RuntimeError(
```
Comment (tjtanaa) [inline vllm/model_executor/layers/sparse_attn_indexer.py:525]: Right now, DSA is only supported using AITER kernels.

File: vllm/model_executor/layers/mhc.py [hunk 6]
```diff
@@ -61,31 +61,35 @@ def forward_hip(
...
+        #         hc_scale,
+        #         hc_base,
+        #         rms_eps,
+        #         hc_pre_eps,
+        #         hc_sinkhorn_eps,
+        #         hc_post_mult_value,
+        #         sinkhorn_repeat,
+        #     )
+        # else:
+        return mhc_kernels.mhc_pre_torch(
```
Comment (tjtanaa) [inline vllm/model_executor/layers/mhc.py:82]: It is not very slow as right now with graph mode and torch compile, this op is optimized to be better than regular torch.

**Review by zyongye — APPROVED** [review]
